Makefile

```diff
@@ -205,10 +222,31 @@ libllama.so: llama.o ggml.o $(OBJS)
 # Tests
 #
 
-benchmark: examples/benchmark/benchmark-q4_0-matmult.c ggml.o $(OBJS)
-	$(CXX) $(CXXFLAGS) $^ -o benchmark-q4_0-matmult $(LDFLAGS)
-	./benchmark-q4_0-matmult
+TARGETS_CPP += benchmark
+DEPS_benchmark := examples/benchmark/benchmark-q4_0-matmult.c build-info.h ggml.o $(OBJS)
+OUTP_benchmark := benchmark-q4_0-matmult
+EXEC_benchmark := ./benchmark-q4_0-matmult
 
 .PHONY: tests
 tests:
 	bash ./tests/run-tests.sh
+
+#
+# Templates
+#
+
+# C++ template
+# To use this template:
+# 1. Add your target to the TARGETS variable: TARGETS_CPP += target
+# 2. Set target-specific dependencies:        DEPS_target := source1 dependency1 dependency2 ...
+# 3. Optionally, set target-specific output:  OUTP_target := output_name
+# 4. Optionally, set target-specific command: EXEC_target := command
+define template_cpp
+OUTP_$(1) ?= $(1)
+$(1): $$(DEPS_$(1)) $$(OBJS) build-info.h
+	$$(CXX) $$(CXXFLAGS) $$(filter-out build-info.h,$$^) -o $$(OUTP_$(1)) $$(LDFLAGS)
+	$$(if $$(value EXEC_$(1)),$$(EXEC_$(1)))
+endef
```

Comment:
This is too "smart", in my humble opinion.

The Makefile could be much simpler, actually.

```Makefile
VPATH= examples examples/main
%.o: %.c
	$(CC)  $(CPPFLAGS) $(CFLAGS)   -c $< -o $@
%.o: %.cpp
	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -c $< -o $@
%: %.o
	$(CXX) $(LDFLAGS) -o $@ $^ $(LDLIBS)
ggml.o: ggml.h ggml-cuda.h
llama.o: ggml.h ggml-cuda.h llama.h llama_util.h
common.o: common.h
main: ggml.o llama.o common.o $(OBJS)
```

Reply:
The Makefile was already pretty busy when I started trying to edit `build-info.h` in. Each target had its own rule. Since the build info has to be generated from git, the example projects have to depend on it. But putting it in the prerequisites was somehow causing the compilers on macOS to try compiling it and failing, so you then have to exclude it from the compiler with `$(filter-out build-info.h, $^)`. I couldn't get it to pass the build tests any other way, so I had to edit that into every example target.

Originally, I had thought I could use a static pattern rule for this (even mentioned as much earlier) and hoped to wrap up all the example target rules into one pattern rule so everyone wouldn't have to see a dozen lines of ` $(filter-out build-info.h, $^)`.

My goal was to put keep all the weird stuff in one place. This way, no has to understand, or copy and paste without understanding, code to change the makefile when a new project is added. There were so many little things different. It seemed each target had its own exception. (main wants to echo text when it's done, benchmark's source is actually `benchmark-q4_0-matmult.c` not `benchmark.c`.

Maybe if I had started with a Makefile from scratch I could have come up with a better solution, but I just wanted to get the example projects in line with the change. I do think a pattern rule is ideal, but I just couldn't get one to work in practice.

I hated the `foreach` eval when I first thought about it, but after stumbling over patterns for way too long, it seemed like the only solution where each example, each program, each test, could just be written the same way in the Makefile. My argument for this approach is it's easy to add future examples and if you need to change the way we're building those examples you just need to edit one line (the template), which also means escaping your `$` with another `$`.

What I can do is revert it to how it was before and just make the bare minimum changes to each target.. After spending a little too much time fighting with patterns, I kind of wish I had just done to begin with, but since I arrived at a destination that wasn't too bad I decided to push it and see what people's thoughts were. Part of the reason I pointed out the "smart" code in a comment after my commit was so that I could try to judge people's comfort level with it. I just want to make sure the code is easy to follow and that people are comfortable updating it if needed.

Reply:
I guess it's not too crazy if I look at the whole file.

> The Makefile was already pretty busy when I started trying to edit build-info.h in. Each target had its own rule. Since the build info has to be generated from git, the example projects have to depend on it. But putting it in the prerequisites was somehow causing the compilers on macOS to try compiling it and failing

Wouldn't it be better to generate `build-info.c` that has global variables and have everyone link to `build-info.o`? It also means that it wouldn't be necessary to rebuild everything on every `git pull` if some files have not changed.

> main wants to echo text when it's done

Double colon rule could work?

> benchmark's source is actually benchmark-q4_0-matmult.c not benchmark.c

I would say it's two targets, a PHONY `benchmark` that depends on `benchmark-q4_0-matmult` and runs it.

Reply:
I do think the general consensus has been that Makefile needs some refactoring, but this PR was never intended to do that. I've reset it to a more minimally changed version and just tried to make it less ugly with `*.h` instead of `build-info.h`, which actually covers future cases of header only includes. I still have my new Makefile stashed away; maybe I'll open a different pull request with a draft for cleaning up Makefile to see if there's any interest there, but I'm less comfortable taking that role as I'm not the person who edits new stuff in regularly.

Building a `build-info.o` would add a little more complexity to the process, but I do agree it would be most effective at preventing unnecessary recompilation. I imagine when there's an official version system, that's how it will be done (but included in `ggml.o`, `llama.o`, or both). I still like the minimalism of the header to get build info into the examples. And while it can cause unnecessary rebuilds, this is limited to the examples and doesn't include the larger files like ggml.c and llama.c and the other `.o` (and `.so`) files.